flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/DecimalType.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.types.logical;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.api.ValidationException;
+
+import java.math.BigDecimal;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Logical type of a decimal number with fixed precision and scale.
+ *
+ * <p>The serialized string representation is {@code DECIMAL(p, s)} where {@code p} is the number of
+ * digits in a number (=precision) and {@code s} is the number of digits to the right of the decimal
+ * point in a number (=scale). {@code p} must have a value between 1 and 38 (both inclusive). {@code s}
+ * must have a value between 0 and {@code p} (both inclusive). The default value for {@code p} is 10.
+ * The default value for {@code s} is 0.
+ */
+@PublicEvolving
+public final class DecimalType extends LogicalType {
+
+	private static final int MIN_PRECISION = 1;
+
+	private static final int MAX_PRECISION = 38;
+
+	private static final int DEFAULT_PRECISION = 10;
+
+	private static final int MIN_SCALE = 0;
+
+	private static final int DEFAULT_SCALE = 0;
+
+	private static final String DEFAULT_FORMAT = "DECIMAL(%d, %d)";
+
+	private static final Set<String> INPUT_OUTPUT_CONVERSION = conversionSet(
+		BigDecimal.class.getName(),
+		"org.apache.flink.table.dataformat.Decimal");
+
+	private static final Class<?> DEFAULT_CONVERSION = BigDecimal.class;
+
+	private final int precision;
+
+	private final int scale;
+
+	public DecimalType(boolean isNullable, int precision, int scale) {
+		super(isNullable, LogicalTypeRoot.DECIMAL);
+		if (precision < MIN_PRECISION || precision > MAX_PRECISION) {
+			throw new ValidationException(
+				String.format(
+					"Decimal precision must be between %d and %d (both inclusive).",
+					MIN_PRECISION,
+					MAX_PRECISION));
+		}
+		if (scale < MIN_SCALE || scale > precision) {
+			throw new ValidationException(
+				String.format(
+					"Decimal scale must be between %d and %d (both inclusive).",
```

Comment:
nit: `"Decimal scale must be between %d and %d (both inclusive)."` -> `"Decimal scale must be between %d and %d(precision) (both inclusive)."`